### PR TITLE
Glob param type should not include ellipsis in extracted variable name

### PR DESCRIPTION
**Pull request by mojombo**
The current implementation of the Glob type in the router includes the `...` in the name of the prop that is passed down to the Page component. The expectation is that a route path like `/path/{path...}` would result in a prop called `path` not `path...`.

This PR also refactors the `paramsForRoute` function to have a more consistent and useful result that cleans up the `matchPath` code.

I've also added some more tests to make sure we're handling all the expected cases properly.

cc @macovedj since this affects the code you wrote recently.

### CODE REVIEW

File: packages/router/src/util.ts [hunk 1]
```diff
@@ -8,22 +8,40 @@ const createNamedContext = <T>(name: string, defaultValue?: T) => {
...
  */
-const paramsForRoute = (route: string) => {
+export const paramsForRoute = (route: string) => {
   // Match the strings between `{` and `}`.
   const params = [...route.matchAll(/\{([^}]+)\}/g)]
 
   return params
     .map((match) => match[1])
     .map((match) => {
-      return match.slice(-3) === '...' ? [match, 'Glob'] : match.split(':')
```
Comment (Tobbe) [inline packages/router/src/util.ts:26]: Changing this to return `[match.slice(0, -3), 'Glob']` for the glob case would have been enough to fix the behavior, right?

File: packages/router/src/__tests__/util.test.ts [hunk 2]
```diff
@@ -121,12 +132,45 @@ describe('matchPath', () => {
...
+    expect(matchPath('/a/{a...}/b/{b...}/c', '/a/1/2/b/3/4/c')).toEqual({
+      match: true,
+      params: {
+        a: '1/2',
+        b: '3/4',
+      },
+    })
+
+    // adjacent
+    expect(matchPath('/a/{a...}{b...}/c', '/a/1/2/3/4/c')).toEqual({
```
Comment (Tobbe) [inline packages/router/src/__tests__/util.test.ts:153]: Did you leave out the slash between `{a...}` and `{b...}` on purpose?
If you did, what would be the expected behavior *with* the slash?

File: packages/router/src/util.ts [hunk 3]
```diff
@@ -8,22 +8,40 @@ const createNamedContext = <T>(name: string, defaultValue?: T) => {
 }
 
 /**
- * Get param name and type transform for a route
+ * Get param name, type, and slug for a route.
```
Comment (Tobbe) [inline packages/router/src/util.ts:11]: You're the master of naming things. But personally I wouldn't call the match a "slug"

image

File: packages/router/src/util.ts [hunk 4]
```diff
@@ -101,19 +105,17 @@ const matchPath = (
   const allParamTypes = { ...coreParamTypes, ...paramTypes }
   let typeConstrainedRoute = route
 
-  // Map all params from the route to their type constraint regex to create a "type-constrained route" regexp
-  for (const [name, type] of routeParams) {
+  // Map all params from the route to their type constraint regex to create a
+  // "type-constrained route" regexp
+  for (const [_name, type, slug] of routeParams) {
```
Comment (Tobbe) [inline packages/router/src/util.ts:110]: See comment about "slug" further up

File: packages/router/src/util.ts [hunk 5]
```diff
@@ -101,19 +105,17 @@ const matchPath = (
   const allParamTypes = { ...coreParamTypes, ...paramTypes }
   let typeConstrainedRoute = route
 
-  // Map all params from the route to their type constraint regex to create a "type-constrained route" regexp
-  for (const [name, type] of routeParams) {
+  // Map all params from the route to their type constraint regex to create a
+  // "type-constrained route" regexp
```
Comment (Tobbe) [inline packages/router/src/util.ts:109]: /regexp?/

<sub>(There's one "regex" and one "regexp" in the comment)</sub>

**Review by Tobbe — COMMENTED** [review]
I think there is a much simpler fix for the ... issue. But I'm guessing you had something more in mind when you made all your changes. Right?

File: packages/router/src/util.ts (hunk 1 above)
Comment (mojombo) [reply packages/router/src/util.ts:26]: Well, almost. There was another place that it had to be tweaked to do the replacement properly, and at that point I wanted to simplify the code, so I did!

File: packages/router/src/__tests__/util.test.ts (hunk 2 above)
Comment (mojombo) [reply packages/router/src/__tests__/util.test.ts:153]: Yes, I wanted to test what happens if you DO happen to have two globs next to each other. It's a fine question about what would happen if there is a slash in there, I will try it out!

File: packages/router/src/util.ts (hunk 3 above)
Comment (mojombo) [reply packages/router/src/util.ts:11]: Fair point, I suppose it can just be called `match`. I will make the change.

File: packages/router/src/util.ts (hunk 5 above)
Comment (mojombo) [reply packages/router/src/util.ts:109]: I will randomly pick one!

File: packages/router/src/__tests__/util.test.ts (hunk 2 above)
Comment (mojombo) [reply packages/router/src/__tests__/util.test.ts:153]: Added a test with a slash between. The first glob matches as much as it can and the 2nd grabs the last bit, which is as expected.

Comment (macovedj) [reply packages/router/src/__tests__/util.test.ts:153]: Oh interesting edge case.  Makes me curious about globs that start and end in slashes as well.  You've likely already thought about it and fiddled with it... but if not I'm curious if that has any impact.

**Comment by macovedj** [comment]
Thanks for cleaning this up all.  These changes definitely seem like improvements.

**Review by Tobbe — APPROVED** [review]
